packages/client/public/root-cookie-accessor-template.html

```diff
@@ -1,141 +1,169 @@
+                    parent.postMessage(JSON.stringify({ hasStorageAccess: false, cookieSet }), e.origin)
+                    break
+                  } else if (storageAccessPermission.state === 'denied') {
+                    message.hasStorageAccess = false
+                    message.cookieSet = cookieSet
+                    message.storageAccessPermission = 'denied'
+                    parent.postMessage(message, e.origin)
+                    break
+                  }
                 }
+              }
+              break
             }
         }
-        document.addEventListener('DOMContentLoaded', async (event) => {
-            document.getElementById('confirmButton').onclick = async event => {
-                try {
-                    const requestStorageAccessResponse = await document.requestStorageAccess()
-                    if (requestStorageAccessResponse == null) {
-                        const cookieSet = document.cookie && document.cookie
-                            .split("; ")
-                            .find((row) => row.startsWith(`visited=`) || row.startsWith('allowedDomains='))
-                        parent.postMessage(JSON.stringify({hasStorageAccess: true, cookieSet}), origin)
-                    }
-                } catch(err) {
-                    if (err.message === 'requestStorageAccess not allowed') parent.postMessage(JSON.stringify({ hasStorageAccess: false, storageAccessPermission: 'denied'}), origin)
-                }
+      }
+      window.onmessage = async function (e) {
+        let allowedDomains
+        try {
+          await document.requestStorageAccess()
+          const allowedDomainCookie = document.cookie.split('; ').find((row) => row.startsWith(`allowedDomains=`))
+          allowedDomains = allowedDomainCookie ? JSON.parse(allowedDomainCookie.split('=')[1]) : []
+        } catch (err) {
+          allowedDomains = []
+        }
+        if (allowedDomains?.indexOf(e.origin) > -1) await processMessage(e)
+        else {
+          const response = await fetch(`<API_URL>/allowed-domains?domainToCheck=${e.origin}`)
+          if (response.status === 200) {
+            const body = await response.json()
+            if (body) {
+              allowedDomains.push(e.origin)
+              document.cookie = `allowedDomains=${JSON.stringify(allowedDomains)}; SameSite=None; Secure`
+              await processMessage(e)
+            }
+          } else {
+            parent.postMessage({ source: source, invalidDomain: true }, e.origin)
+          }
+        }
+      }
+      document.addEventListener('DOMContentLoaded', async (event) => {
+        document.getElementById('confirmButton').onclick = async (event) => {
+          try {
+            const requestStorageAccessResponse = await document.requestStorageAccess()
+            if (requestStorageAccessResponse == null) {
+              const cookieSet =
+                document.cookie &&
+                document.cookie
```

Comment:
Document.cookie is being processed multiple times all over the place, creating and allocating in memory each time a new collection (split), iterating multiple times (find), mutating iterations (replace), etc.

As a general rule, each time we realize that we are doing the same operation more than once, we must reconsider an alternate approach.

The best way to avoid iterating the same data over and over is by creating a key-value map once, enabling data values to be accessible without the need of iterating the collection each time that a value is required.

```suggestion
   // We iterate just once and allocate in memory once
   const cookiesMapObj = document.cookies.split(';').reduce((accumulator, current) => {
  if (!accumulator) accumulator = {}
  const values = current.split('=')
  return {...accumulator, value[0]: value[1]}
})
```

Now each time we need to access a cookie data value we can simply do it by the key reference

```suggestion
 someFunction(cookiesMapObj['user-token'])
```

Now we avoided tons of iterations, duplicated code, unnecessary complexity, misused memory allocation and a it's lot less prone to failure.